articles/logging-with-winston/index.md

```diff
@@ -0,0 +1,547 @@
+---
+layout: engineering-education
+status: publish
+published: true
+url: /engineering-education/logging-with-winston/
+title: 
+description: This article goes through 
+author: joseph-chege
+date: 2020-11-10T00:00:00-15:00
+topics: []
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/logging-with-winston/hero.png
+    alt:  example image
+---
+Assume you have an app running at the production level. If this app runs into an error mode, a million things could lead to that. Some are essential others are not. As a developer, you need to find the critical issues and solve them.
+<!--more-->
+
+What are the possibilities of solving that problem? Perhaps by going back to the code and check if every line of code runs as expected. Well, you might think that it is possibly easier with a small app. However, there is a high chance this will take you time and energy of concentration trying to pip on your small app codes. You get the idea, how this could be for an extensive system. Suppose, let's say, an instance where the app collects some user's information and save them into a database.
+
+If the system fails, the server will return a system error to the user end. It would be best if you caught those instances and solve such errors. In such a case, how do you get to know that user a or use b run into an individual system error?
+
+If you need to avoid such kind of frustration, you cannot avoid logging. Yes, logging is something you can’t avoid. A log is the first place to look as a programmer, to track down errors and flow of events, especially from a server. A log tells you what happens when an app is running and interacting with your users. A great use case for logging would be if, for example, you have a bug in your system, and you want to understand the steps that led up to its occurrence.
+
+Logging is a process of recording information generated by application activities into log files. Messages saved in the log file are called logs. A log is a single instance recorded in the log file.
+
+Structuring application logging in Node.js is critical. When running an app locally, you can hook it to a debugger, which is excellent, to find out problems when running the app. During development, you would typically use `console.log` to get the application logs. But when an application goes to production and the user starts to interact with it, you can’t `console.log` any more. If something went wrong and the app crashed, you cannot check that with the console. It would help if you had a concise, clean, and quality logging middleware such as [Winston](https://github.com/winstonjs/winston). Winston process your app activities and generate useful information into a log file or a database. You can afterwards check every activity generated by your application.
```

Comment:
It would be nice if you explained to users what a debugger is or linked to a source that does.